buffer.c

```diff
@@ -282,56 +201,67 @@ PHP_METHOD(ArrayBuffer, unserialize)
 	p = (unsigned char *) str;
 	max = (unsigned char *) str + str_len;
 
-	INIT_ZVAL(zv);
-	if (!php_var_unserialize(&zv_ptr, &p, max, &var_hash TSRMLS_CC)
-	    || Z_TYPE_P(zv_ptr) != IS_STRING || Z_STRLEN_P(zv_ptr) == 0) {
-		zend_throw_exception(NULL, "Could not unserialize buffer", 0 TSRMLS_CC);
+	zbuf = var_tmp_var(&var_hash);
+	if (!php_var_unserialize(zbuf, &p, max, &var_hash)) {
+		zend_throw_exception(NULL, "Could not unserialize buffer: no data", 0);
 		goto exit;
-	}
-
-	intern->buffer = Z_STRVAL_P(zv_ptr);
-	intern->length = Z_STRLEN_P(zv_ptr);
+        }
+        if(Z_TYPE_P(zbuf) != IS_STRING) {
+		zend_throw_exception(NULL, "Could not unserialize buffer: not a string", 0);
+		goto exit;
+       }
 
-	INIT_ZVAL(zv);
-	if (!php_var_unserialize(&zv_ptr, &p, max, &var_hash TSRMLS_CC)
-	    || Z_TYPE_P(zv_ptr) != IS_ARRAY) {
-		zend_throw_exception(NULL, "Could not unserialize properties", 0 TSRMLS_CC);
+        zstr = zval_get_string(zbuf);
+	intern->length = ZSTR_LEN(zstr);
+        if(intern->length == 0) {
+		zend_throw_exception(NULL, "Could not unserialize buffer: empty string", 0);
+		goto exit;
+	}
+	intern->buffer = emalloc(intern->length);
+        memcpy(intern->buffer, &ZSTR_VAL(zstr), intern->length);
+        zend_string_release(zstr);
+
+	ztable = var_tmp_var(&var_hash);
+	if (!php_var_unserialize(ztable, &p, max, &var_hash)
+			|| Z_TYPE_P(ztable) != IS_ARRAY) {
+		zend_throw_exception(NULL, "Could not unserialize properties", 0);
 		goto exit;
 	}
 
-	if (zend_hash_num_elements(Z_ARRVAL_P(zv_ptr)) != 0) {
+	if (zend_hash_num_elements(Z_ARRVAL_P(ztable)) != 0) {
 		zend_hash_copy(
-			zend_std_get_properties(getThis() TSRMLS_CC), Z_ARRVAL_P(zv_ptr),
-			(copy_ctor_func_t) zval_add_ref, NULL, sizeof(zval *)
+			zend_std_get_properties(getThis()), Z_ARRVAL_P(ztable),
+			(copy_ctor_func_t) zval_add_ref
 		);
 	}
 
 exit:
-	zval_dtor(zv_ptr);
 	PHP_VAR_UNSERIALIZE_DESTROY(var_hash);
 }
 
-static void array_buffer_view_free_object_storage(buffer_view_object *intern TSRMLS_DC)
+static void array_buffer_view_free(zend_object *obj)
 {
-	zend_object_std_dtor(&intern->std TSRMLS_CC);
-
-	if (intern->buffer_zval) {
-		zval_ptr_dtor(&intern->buffer_zval);
+	buffer_view_object *intern = php_buffer_view_fetch_object(obj);
+	if(!intern){
+			    return;
 	}
 
+	zend_object_std_dtor(&intern->std);
 	efree(intern);
```

Comment:
efree needs to be dropped.

Reply:
Thanks. The free'ing and the destructor gave me a lot of grief but I had hoped the tests should have caught this up. I wonder what type of testing would have found this.

Reply:
Either valgrind or address sanitizer can be used to catch such issues. Valgrind is slow but probably easier for extensions. (Unless doing an in-tree build, in which case `--enable-address-sanitizer` can be used.)